### ISSUE TITLE
product-find fails to infer properly

### ISSUE THREAD

**Issue by bpollack** [factor/factor#2381]
While playing with this year's Advent of Code, I hit an inference bug with `product-find`.

(Slight spoilers for day one of the AoC ahead.)

A quick solution that works for the problem is the little snippet `[ { [ all-unique? ] [ sum 2020 = ] 1&& } ] product-find`. This works fine when written inline in the listener, but if you try to define it as an actual word (e.g. `: summing-tuple ( seqs-of-numbers -- tuple ) [ { [ all-unique? ] [ sum 2020 = ] } 1&& ] product-find ;` edit by @erg: fixed `1&& } -> } 1&&`), Factor will error out with a stack effect violation. That's wrong, though: the stack effect is correct.

The issue appears to be a bug in the inference engine. To wit: `[ { [ all-unique? ] [ sum 2020 = ] 1&& } ] infer` properly says `( x -- x )`, but `[ [ { [ all-unique? ] [ sum 2020 = ] } 1&& ] product-find ] infer` fails. The resulting error is hard for me to quickly understand, since it's hitting a lexical word in a pile of `POSTPONE:`s, but is likely worth looking into.

**Comment by bpollack** [comment]
(Doing the slower `<product-sequence> [ ... ] find nip` works fine, incidentally, so this is trivial to work around; just odd.)

**Comment by mrjbq7** [comment]
This infers properly:

```factor
IN: scratchpad [ { [ all-unique? ] [ sum 2020 = ] } 1&& ] infer .
( x -- x )
```

And it looks like it's not an inference bug, but an ``unless`` that didn't leave the stack consistent in the other case... fixed here maybe:

4d946fe8af4dec845a307f299f7b55fe76ce4f59

Give that a try!

**Comment by bpollack** [comment]
Bah, you're entirely correct; I should've just read `product-find` more carefully. Good catch!

**Comment by jonenst** [comment]
This means we need tests for product-find ..?

**Comment by jonenst** [comment]
Also the docs Say:
"[ ... ] product-each is equivalent to, but more efficient than, <product-sequence> [ ... ] find."
Copy paste error from product-each ?

**Comment by jonenst** [comment]
In your initial comment, erg fixed" ] 1&& }" but there are two other places with the mistake

**Comment by mrjbq7** [comment]
Good idea @jonenst docs fixed in 139f6bc26b5400d34d46edd6b9aa5081647cf280 tests added in 1f6d3a11b783033649f42c589b8fd9ddd1cc9981.

**Comment by jonenst** [comment]
👍
